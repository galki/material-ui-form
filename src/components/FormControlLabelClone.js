import React from 'react'
import PropTypes from 'prop-types'
import _ from 'lodash' // eslint-disable-line import/no-extraneous-dependencies

import Checkbox from 'material-ui/Checkbox'
import Switch from 'material-ui/Switch'
import { FormControlLabel } from 'material-ui/Form'


export default class FormControlLabelClone extends React.Component {
  static propTypes = {
    /* eslint-disable-next-line */
    control: PropTypes.object.isRequired,
    field: PropTypes.object,
    label: PropTypes.string.isRequired,
    onToggle: PropTypes.func.isRequired,
    onConstruct: PropTypes.func.isRequired,
  }

  static defaultProps = {
    field: {},
  }

  constructor(props) {
    super(props)

    if (![Checkbox, Switch].includes(props.control.type)) {
      throw new Error('invalid FormControlLabel control component')
    }

    let { checked } = props.control.props
    const { value } = props.control.props

    if (props.field.value === undefined) {
      props.onConstruct(props.control.props)
    } else {
      checked = _.get(props.field, 'checked')
    }

    this.state = {
      checked,
      value,
    }
  }

  onToggle = (event, checked) => {
    checked = _.get(event, 'target.checked') || checked
    let { value } = this.props.control.props // eslint-disable-line react/prop-types
    const { name } = this.props.control.props // eslint-disable-line react/prop-types
    value = checked ? value : ''
    this.setState({ checked, value })
    this.props.onToggle(name, value, checked)
  }

  render() {
    const { control, label } = this.props
    const onChange = (
      control.props.onChange || control.props.onToggle || this.onToggle
    )
    const controlOptions = {
      checked: this.state.checked,
      onChange,
      value: this.state.value,
    }

    return (
      <FormControlLabel
        checked={this.state.checked}
        control={React.cloneElement(control, controlOptions)}
        onChange={onChange}
        label={label}
        value={this.state.value}
      />
    )
  }
}
