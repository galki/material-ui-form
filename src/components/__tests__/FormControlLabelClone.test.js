import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Checkbox from 'material-ui/Checkbox'
import Switch from 'material-ui/Switch'

import FormControlLabelClone from '../FormControlLabelClone'


describe('<FormControlLabelClone>:<Checkbox>', () => {
  const field = {
    isPristine: true,
    isRequired: null,
    pristineValue: null,
    validations: [],
    validators: [],
    value: undefined,
  }

  const wrapper = shallow(
    <FormControlLabelClone
      control={<Checkbox checked={false} name="love" value="yes" />}
      field={field}
      label="I love love"
      onConstruct={jest.fn()}
      onToggle={jest.fn()}
    />
  )

  it('should render', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should not throw if field type prop is Checkbox or Switch', () => {
    function checkType() {
      if (![Checkbox, Switch].includes(wrapper.instance().props.control.type)) {
        throw new Error('FieldClone does not support native elements')
      }
    }
    expect(checkType).not.toThrow()
  })

  it('should call onConstruct', () => {
    expect(wrapper.instance().props.onConstruct).toHaveBeenCalled()
  })

  it('should set state', () => {
    const { control } = wrapper.instance().props
    expect(wrapper.state()).toMatchObject({
      checked: control.props.checked,
      value: control.props.value,
    })
  })

  it('should handle onToggle events', () => {
    const checked = true
    const event = { target: { checked } }
    wrapper.simulate('change', event)
    expect(wrapper.state()).toMatchObject({
      checked,
      value: wrapper.state('value'),
    })
    expect(wrapper.instance().props.onToggle).toHaveBeenCalled()
  })
})
