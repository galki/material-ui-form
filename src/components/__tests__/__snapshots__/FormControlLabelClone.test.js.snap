// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormControlLabelClone>:<Checkbox> should render 1`] = `
<WithStyles(FormControlLabel)
  checked={false}
  control={
    <WithStyles(Checkbox)
      checked={false}
      name="love"
      onChange={[Function]}
      value="yes"
    />
  }
  label="I love love"
  onChange={[Function]}
  value="yes"
/>
`;
